### PR TITLE
Fix bad suggestions for `deref_addrof` and `try_err` lints

### PR DESCRIPTION
**Pull request by ThibsG**
Fix bad suggestions when in macro expansion for `deref_addrof` and `try_err` lints.

Fixes: #6234 
Fixes: #6242 
Fixes: #6237 

changelog: none

r? @llogiq

### CODE REVIEW

File: clippy_lints/src/reference.rs [hunk 1]
```diff
@@ -46,13 +46,19 @@ impl EarlyLintPass for DerefAddrOf {
             if !in_macro(addrof_target.span);
             then {
                 let mut applicability = Applicability::MachineApplicable;
+                let sugg = if e.span.from_expansion() {
+                    let snip = snippet_with_applicability(cx, e.span, "_", &mut applicability);
+                    snip.trim_start_matches(|c| c == '&' || c == '*').to_string()
```
Comment (ThibsG) [inline clippy_lints/src/reference.rs:51]: Seems too hacky, I will try to find something better.

File: clippy_lints/src/reference.rs [hunk 2]
```diff
@@ -42,17 +43,45 @@ impl EarlyLintPass for DerefAddrOf {
...
+
+                        let rpos = if *mutability == Mutability::Mut {
+                            macro_source.rfind("mut").expect("already checked this is a mutable reference") + "mut".len()
+                        } else {
+                            macro_source.rfind('&').expect("already checked this is a reference") + "&".len()
+                        };
+                        #[allow(clippy::cast_possible_truncation)]
+                        let span_after_ref = e.span.with_lo(BytePos(e.span.lo().0 + rpos as u32));
+                        let span = trim_leading_whitespaces(span_after_ref);
+                        snippet_with_applicability(cx, span, "_", &mut applicability).to_string()
```
Comment (llogiq) [inline clippy_lints/src/reference.rs:71]: We could move the `.to_string()` here and in the other 2 branches (line 73 and 76) to either line 77 or 84.

File: clippy_lints/src/try_err.rs [hunk 3]
```diff
@@ -92,7 +92,7 @@ impl<'tcx> LateLintPass<'tcx> for TryErr {
 
                 let expr_err_ty = cx.typeck_results().expr_ty(err_arg);
 
-                let origin_snippet = if err_arg.span.from_expansion() {
+                let origin_snippet = if err_arg.span.from_expansion() && !in_macro(expr.span) {
```
Comment (llogiq) [inline clippy_lints/src/try_err.rs:95]: This change seems unrelated?

**Review by llogiq — COMMENTED** [review]
I only have one nit and one question, otherwise this looks good to go. :+1:

File: clippy_lints/src/try_err.rs (hunk 3 above)
Comment (ThibsG) [reply clippy_lints/src/try_err.rs:95]: This `!in_macro(expr.span)` is here to prevent the use of the macro callsite for suggestion.
In this example:
```
macro_rules! try_validation {
    ($e: expr) => {{
        match $e {
            Ok(_) => 0,
            Err(_) => Err(1)?, // <-- lint here
        }
    }};
}

fn calling_macro() -> Result<i32, i32> {
    try_validation!(Ok::<_, i32>(5));
    Ok(5)
}
```
For information,`err_arg.span` is referring to the `1` from `Err(1)?`
We want the suggestion to be `return Err(1)` and not to be the macro call site (`try_validation!(Ok::<_, i32>(5));`).

**Comment by ThibsG** [comment]
Just saw #6237, it is the same kind of problem we're having here.
If it's ok for you to fix this here, I can also work on this before we merge this.

**Comment by ThibsG** [comment]
Ok now it should be good for #6237 as well

**Comment by llogiq** [comment]
Good work! Thank you. @bors r+
